ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_program_factory.cpp

```diff
@@ -3,352 +3,156 @@
-                                                    input_0_stick_size,
-                                                    input_1_stick_size,
-                                                    input_0_stride,
-                                                    input_1_stride,
-                                                    num_output_rows_per_core * num_input_tensors ,
-                                                    num_pages_per_risc,
-                                                    num_output_rows_per_core,
-                                                    num_pages_per_risc*output_stick_size,
-                                                    num_pages_per_risc*input_0_stick_size,
-                                                    num_pages_per_risc*input_1_stick_size,
-                                                };
-
-
-
-
-
-    tt_metal::KernelHandle unary_reader_kernel_id = tt_metal::CreateKernel(
-        program,
-        "ttnn/cpp/ttnn/operations/data_movement/concat/device/kernels/dataflow/reader_height_sharded_width_concat_two_tensors.cpp",
-        all_cores,
-        tt_metal::ReaderDataMovementConfig(compile_time_args_0));
-
-
-    tt_metal::KernelHandle unary_writer_kernel_id = tt_metal::CreateKernel(
-        program,
-        "ttnn/cpp/ttnn/operations/data_movement/concat/device/kernels/dataflow/reader_height_sharded_width_concat_two_tensors.cpp",
-        all_cores,
-        tt_metal::WriterDataMovementConfig(compile_time_args_1));
-
-
-
-
-    auto override_runtime_arguments_callback =
-        [unary_reader_kernel_id, unary_writer_kernel_id, all_cores, num_input_tensors](
-            const void *operation,
-            Program &program,
-            const std::vector<Tensor> &input_tensors,
-            const std::vector<std::optional<const Tensor>> &,
-            const std::vector<Tensor> &output_tensors) {
-                ;
-        };
-
-    return {.program = std::move(program), .override_runtime_arguments_callback = override_runtime_arguments_callback};
+    return page_size;
 }
 
+}  // namespace
 
+namespace ttnn::operations::data_movement::detail {
 
+// Concat sharded tensors into sharded output in row-major layout. Currently it only supports height-sharded width
+// concat or width-sharded height concat.
+//
+// It is done by copying each row of each input sharded tensor to the right offset in the sharded output tensor based on
+// the sharded output width in bytes (output stride). This way works for both width and height concat.
+//
+// For example in width concat, rows of an input tensor are placed at the same column offset but sequential rows in the
+// output. The memory address gap between neighbor input rows is exactly the output width. In height concat, all input
+// rows are placed at column 0 but sequential rows in the output. The address gap between neighbor input rows is still
+// the output width (which is equal to the input width).
```

Comment:
For now I directly replaced the `s2s_rm_concat_multi_core` and removed `s2s_rm_concat_two_tensors_multi_core` for the two tensors as the new kernel should support the cases they covered.

Let me know if we want to put this new kernel as experimental separately and maybe guarded by a flag. Or if there is a special reason we want to keep `s2s_rm_concat_two_tensors_multi_core`

Reply:
ya two tensors multi core is just a manually unrolled version of the general case (it should be covered by it).

but we had a model where concat was the bottleneck and using runtime args (which is how we added variadic number of inputs) added a bunch of cycles. I think for now guarding against that with a flag is okay (if input is 2 tensors ) . 

I know @TT-billteng is looking for efficient ways to send variadic args in compile-time args and after that we should be able to remove the manually unrolled kernel.

Reply:
SG. I restored the original two tensors concat kernel. If we want to also support width-sharded hight concat with two tensors in the meantime, I'm happy to open another PR to replace `s2s_rm_concat_two_tensors_multi_core` with an unrolled version of this new kernel

Reply:
Regarding to using a flag to guard new kernel for the input is 2 tensors, I'm actually not clear about how the flag is used in tt-metal library. So for now I just restore the `s2s_rm_concat_two_tensors_multi_core` to keep it simple.

Let me know if we really want a flag and if there is an example on how the flag is used